Optimise refreshes and updates of elevation profile charts

nyalldawson opened the pull request:
This PR improves the handling of updates for elevation profile charts, so that:

- we only redraw existing results if a chart presentation only setting is changed (e,g the line color), instead of refetching all the results for that layer
- we don't refetch results for ALL layers when a elevation property (such as layer extrusion) is changed, just the affected layer
- charts are automatically updated whenever things change, including automatic updates when vector features are modified

(This is also a step towards progressive "refinement" of profiles, where for raster/point cloud layers we'll refetch the results as the plot is zoomed in/out, using an appropriate level of detail for that chart zoom range.)

nyalldawson committed:
Add signals for when layer elevation properties change

nyalldawson committed:
Add means of linking profile results back to source layer

nyalldawson committed:
Fix clazy warning

nyalldawson committed:
Add api to trigger regeneration of profile results for a single layer

nyalldawson committed:
Automatically update plot results when associated layer elevation
properties are changed

nyalldawson committed:
Also automatically update elevation profile plots when layer data
changes

Eg. when vector features are modified the associated layer should
be updated in the plot immediately

nyalldawson committed:
Add api to allow redrawing of previously collected results for a single
elevation profile source

And use this to just redraw the existing results whenever a setting
relating only to the appearance of the profile chart is changed,
instead of regenerating the whole chart

wonder-sk commented on src/gui/elevation/qgselevationprofilecanvas.cpp:
```diff
@@ -547,7 +638,94 @@ void QgsElevationProfileCanvas::generationFinished()
+{
+  mPlotItem->update();
+  mDeferredRedrawScheduled = false;
+}
```
I have been getting a bit confused what's the difference between an "update" and a "redraw" - maybe worth commenting somewhere which does what...

nyalldawson replied:
Ok, I've renamed all the references to "update" meaning "rebuild the profile for a layer" to "regenerate", so that we don't get the naming clash with Qt's "update=repaint" association.

wonder-sk commented on src/core/elevation/qgsabstractprofilegenerator.h:
```diff
@@ -198,6 +209,11 @@ class CORE_EXPORT QgsAbstractProfileGenerator
+    /**
+     * Returns a unique identifier representing the source of the profile.
+     */
+    virtual QString sourceId() const = 0;
```
It looks like we're using layer IDs for the time being. Do we want to comment that those are normally layer IDs, or better to keep the IDs opaque? (i.e. whether anyone could ever rely on converting source ID to layer)

nyalldawson replied:
Ok, added to the docs. And yes, they'll always match layer IDs for generators associated with a map layer. Other custom generators (e.g. ones from plugins) will have their own interpretation.

wonder-sk commented on src/core/elevation/qgsabstractprofilegenerator.h:
```diff
@@ -164,6 +165,16 @@ class CORE_EXPORT QgsAbstractProfileResults
+     *
+     * The base class method does nothing.
+     */
+    virtual void updateFromGenerator( const QgsAbstractProfileGenerator *generator );
```
I am quite unsure about semantics of this call... Right now all implementations just update the appearance-related pieces of data. But the method doxygen description (first sentence) made me think that in an implementation also results data should be overwritten from the given generator.

I wonder if the method could be renamed somehow to make it more clear we're only after copying generator properties...

nyalldawson replied:
I've renamed to "copyPropertiesFromGenerator" and added clarification to the documentation

wonder-sk reviewed (approved):
Good stuff!

nyalldawson committed:
Add dox note about source ID and layer IDs

nyalldawson committed:
Rename method to copyPropertiesFromGenerator and avoid some duplicate code

nyalldawson committed:
Use "regenerate" instead of "update" for re-building the profile for
a layer, as "update" is ambiguous and is associated with "repainting"
a Qt widget
